File: backend/services/runner/src/shared/filereview/__tests__/digest.test.ts
/**
 * The TypeScript driver of the cross-edition file-digest corpus
 * (`apis/testdata/hitl/file-digest/vectors.json`). The Go (`digest_test.go`) and
 * Java (`FileDigestTest`) suites replay the SAME vectors, so a passing run in all
 * three proves the runner computes byte-identical `file_digest` /
 * `aggregate_digest` to both backends — the load-bearing parity contract for the
 * producer.
 */

import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { existsSync } from "node:fs";
import { describe, expect, it } from "vitest";
import { FileChangeKind } from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/enum_pb";
import { aggregateDigest, fileChangeKindName, fileDigest, type FileDigestInput } from "../digest.js";

interface VectorCase {
  name: string;
  path_before: string;
  path_after: string;
  kind: string;
  before_sha256: string;
  after_sha256: string;
  file_digest: string;
}

interface VectorAggregate {
  name: string;
  change_names: string[];
  aggregate_digest: string;
}

interface Vectors {
  cases: VectorCase[];
  aggregates: VectorAggregate[];
}

/** Walk up from this file until the shared HITL fixtures are found. */
function findVectorsFile(): string {
  let dir = dirname(fileURLToPath(import.meta.url));
  for (let i = 0; i < 12; i++) {
    const candidate = join(dir, "apis", "testdata", "hitl", "file-digest", "vectors.json");
    if (existsSync(candidate)) return candidate;
    dir = dirname(dir);
  }
  throw new Error("could not locate apis/testdata/hitl/file-digest/vectors.json");
}

/** Map the proto enum value name in the corpus back to the TS enum member. */
function kindFromName(name: string): FileChangeKind {
  for (const k of [
    FileChangeKind.UNSPECIFIED,
    FileChangeKind.ADD,
    FileChangeKind.MODIFY,
    FileChangeKind.DELETE,
    FileChangeKind.RENAME,
    FileChangeKind.BINARY_CHANGE,
  ]) {
    if (fileChangeKindName(k) === name) return k;
  }
  throw new Error(`unknown FileChangeKind name in corpus: ${name}`);
}

const vectors: Vectors = JSON.parse(readFileSync(findVectorsFile(), "utf8"));

function toInput(c: VectorCase): FileDigestInput {
  return {
    pathBefore: c.path_before,
    pathAfter: c.path_after,
    kind: kindFromName(c.kind),
    beforeSha256: c.before_sha256,
    afterSha256: c.after_sha256,
  };
}

describe("file-digest cross-edition corpus", () => {
  it("loads the shared corpus (guard against an empty/missing file)", () => {
    expect(vectors.cases.length).toBeGreaterThanOrEqual(4);
    expect(vectors.aggregates.length).toBeGreaterThanOrEqual(3);
  });

  for (const c of vectors.cases) {
    it(`file_digest matches for "${c.name}"`, () => {
      expect(fileDigest(toInput(c))).toBe(c.file_digest);
    });
  }

  const byName = new Map(vectors.cases.map((c) => [c.name, c]));
  for (const agg of vectors.aggregates) {
    it(`aggregate_digest matches for "${agg.name}"`, () => {
      const inputs = agg.change_names.map((n) => {
        const c = byName.get(n);
        if (!c) throw new Error(`aggregate references unknown case ${n}`);
        return toInput(c);
      });
      expect(aggregateDigest(inputs)).toBe(agg.aggregate_digest);
    });
  }
});

File: backend/services/stigmer-server/pkg/domain/agentexecution/filereview/digest_test.go
package filereview

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	agentexecutionv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/agentexecution/v1"
)

// TestFileDigestVectors locks the canonical file_digest / aggregate_digest
// functions against apis/testdata/hitl/file-digest/vectors.json — expected values
// computed independently (plain sha256), so a passing run here AND in the Java
// mirror proves cross-edition parity, not self-consistency.
func TestFileDigestVectors(t *testing.T) {
	corpus := loadDigestCorpus(t)

	if len(corpus.Cases) < 4 {
		t.Fatalf("loaded %d digest cases, want >= 4 (corpus missing or not discovered)", len(corpus.Cases))
	}

	byName := make(map[string]*agentexecutionv1.CapturedFileChange, len(corpus.Cases))
	for _, c := range corpus.Cases {
		kindVal, ok := agentexecutionv1.FileChangeKind_value[c.Kind]
		if !ok {
			t.Fatalf("case %q: unknown FileChangeKind %q", c.Name, c.Kind)
		}
		change := &agentexecutionv1.CapturedFileChange{
			PathBefore:   c.PathBefore,
			PathAfter:    c.PathAfter,
			Kind:         agentexecutionv1.FileChangeKind(kindVal),
			BeforeSha256: c.BeforeSha256,
			AfterSha256:  c.AfterSha256,
		}
		byName[c.Name] = change
		if got := FileDigest(change); got != c.FileDigest {
			t.Errorf("FileDigest(%s) = %s, want %s", c.Name, got, c.FileDigest)
		}
	}

	for _, agg := range corpus.Aggregates {
		changes := make([]*agentexecutionv1.CapturedFileChange, 0, len(agg.ChangeNames))
		for _, n := range agg.ChangeNames {
			change, ok := byName[n]
			if !ok {
				t.Fatalf("aggregate %q references unknown case %q", agg.Name, n)
			}
			changes = append(changes, change)
		}
		if got := AggregateDigest(changes); got != agg.AggregateDigest {
			t.Errorf("AggregateDigest(%s) = %s, want %s", agg.Name, got, agg.AggregateDigest)
		}
	}
}

type digestCorpus struct {
	Cases []struct {
		Name         string `json:"name"`
		PathBefore   string `json:"path_before"`
		PathAfter    string `json:"path_after"`
		Kind         string `json:"kind"`
		BeforeSha256 string `json:"before_sha256"`
		AfterSha256  string `json:"after_sha256"`
		FileDigest   string `json:"file_digest"`
	} `json:"cases"`
	Aggregates []struct {
		Name            string   `json:"name"`
		ChangeNames     []string `json:"change_names"`
		AggregateDigest string   `json:"aggregate_digest"`
	} `json:"aggregates"`
}

func loadDigestCorpus(t *testing.T) digestCorpus {
	t.Helper()
	path := filepath.Join(repoTestdataHitl(t), "file-digest", "vectors.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("reading digest vectors %s: %v", path, err)
	}
	var corpus digestCorpus
	if err := json.Unmarshal(raw, &corpus); err != nil {
		t.Fatalf("decoding digest vectors %s: %v", path, err)
	}
	return corpus
}
